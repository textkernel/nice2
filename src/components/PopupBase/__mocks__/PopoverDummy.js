import * as React from 'react';
import PropTypes from 'prop-types';
import { Button } from '../../..';

const PopoverDummy = React.forwardRef((props, ref) => {
    const { setPopupVisibility } = props;
    const popupStyle = {
        backgroundColor: '#FFEEDD',
        padding: '10px',
        borderRadius: '5px',
        border: '1px solid #CCC',
        boxShadow: '0 3px 6px 6px rgba(0, 0, 0, 0.05)',
    };
    return (
        <div style={popupStyle} ref={ref}>
            <p>I am the content of pop up.</p>
            <p>Can be a form or anything.</p>
            <Button onClick={() => setPopupVisibility(false)}>Close me!</Button>
        </div>
    );
});

PopoverDummy.propTypes = {
    setPopupVisibility: PropTypes.func.isRequired,
};

PopoverDummy.displayName = 'Popover';

export default PopoverDummy;
