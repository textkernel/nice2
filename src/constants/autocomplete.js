/* eslint-disable import/prefer-default-export */
export const NUMBER_OF_SUGGESTION_LOADING_PLACEHOLDERS = 5;
