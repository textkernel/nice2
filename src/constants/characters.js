// eslint-disable-next-line import/prefer-default-export
export const CROSS_CHAR = '\u00D7';
