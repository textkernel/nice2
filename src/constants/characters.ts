export const CROSS_CHAR = '\u00D7';
