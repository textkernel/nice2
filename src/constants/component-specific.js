/** Constants that are component specific but do not merit a separate file */

/** Autocomplete */
export const NUMBER_OF_SUGGESTION_LOADING_PLACEHOLDERS = 5;

/** Input */
export const INPUT_TYPES = [
    'text',
    'password',
    'email',
    'date',
    'datetime-local',
    'month',
    'tel',
    'number',
    'url',
    'search',
    'time',
    'week',
];

/** List */
export const LIST_NAVIGATION_DIRECTIONS = {
    UP: 'ArrowUp',
    DOWN: 'ArrowDown',
};

/** PopupBase */
export const POPUP_PLACEMENTS = [
    'auto',
    'auto-start',
    'auto-end',
    'top',
    'top-start',
    'top-end',
    'right',
    'right-start',
    'right-end',
    'bottom',
    'bottom-start',
    'bottom-end',
    'left',
    'left-start',
    'left-end',
];
