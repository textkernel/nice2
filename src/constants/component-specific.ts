/** Constants that are component specific but do not merit a separate file */

/** Autocomplete */
export const NUMBER_OF_SUGGESTION_LOADING_PLACEHOLDERS = 5;

/** Button */
export const BUTTON_TYPES = ['submit', 'button'] as const;

export type ButtonTypes = typeof BUTTON_TYPES[number];

/** Input */
export const INPUT_TYPES = [
    'text',
    'password',
    'email',
    'date',
    'datetime-local',
    'month',
    'tel',
    'number',
    'url',
    'search',
    'time',
    'week',
] as const;

export type InputTypes = typeof INPUT_TYPES[number];

/** List */
export const LIST_NAVIGATION_DIRECTIONS = {
    UP: 'ArrowUp',
    DOWN: 'ArrowDown',
};

/** PopupBase */
export const POPUP_PLACEMENTS = [
    'auto',
    'auto-start',
    'auto-end',
    'top',
    'top-start',
    'top-end',
    'right',
    'right-start',
    'right-end',
    'bottom',
    'bottom-start',
    'bottom-end',
    'left',
    'left-start',
    'left-end',
];
