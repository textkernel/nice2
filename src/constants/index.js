export const INPUT_TYPES = [
    'text',
    'password',
    'email',
    'date',
    'datetime-local',
    'month',
    'tel',
    'number',
    'url',
    'search',
    'time',
    'week',
];

export const LIST_NAVIGATION_DIRECTIONS = {
    UP: 'ArrowUp',
    DOWN: 'ArrowDown',
};

export const CROSS_CHAR = 'Ã—';

export * from './style-related';
export * from './autocomplete';
export * from './keyboard';
