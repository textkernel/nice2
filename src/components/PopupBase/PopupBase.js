import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import PopperJS from 'popper.js';
import { POPUP_PLACEMENTS, ESCAPE_KEY } from '../../constants';

class PopupBase extends React.Component {
    static checkDOMPath(domElem, checkIsTarget) {
        let currentElem = domElem;
        while (currentElem && currentElem !== document) {
            if (checkIsTarget(currentElem)) return true;
            currentElem = currentElem.parentElement;
        }
        return false;
    }

    constructor(props) {
        super(props);
        const { anchorRef, popupRef } = props;

        this.state = { isOpened: false };
        this.popper = undefined;
        this.anchorRef = anchorRef || React.createRef();
        this.popupRef = popupRef || React.createRef();

        this.setPopupVisibility = this.setPopupVisibility.bind(this);
        this.handleWindowClick = this.handleWindowClick.bind(this);
        this.handleKeyPress = this.handleKeyPress.bind(this);
    }

    componentDidMount() {
        document.addEventListener('click', this.handleWindowClick);
        document.addEventListener('keydown', this.handleKeyPress, false);
    }

    componentDidUpdate(prevProps, prevState) {
        const { isOpened } = this.state;
        if (isOpened && !prevState.isOpened) {
            this.createPopperInstance();
        }
        if (!isOpened && prevState.isOpened) {
            this.destroyPopperIstance();
        }
    }

    componentWillUnmount() {
        document.removeEventListener('click', this.handleWindowClick);
        document.removeEventListener('keydown', this.handleKeyPress, false);
    }

    setPopupVisibility(shouldBeOpen) {
        const { isOpened } = this.state;
        if (shouldBeOpen !== isOpened) {
            this.setState({ isOpened: shouldBeOpen });
        }
    }

    getArgs() {
        const { isOpened } = this.state;
        return {
            setPopupVisibility: this.setPopupVisibility,
            isOpened,
        };
    }

    handleWindowClick(event) {
        if (!this.popupRef.current) return;
        const wasPopupClicked = PopupBase.checkDOMPath(
            event.target,
            elem => elem.dataset && !!elem.dataset.popup
        );
        const wasOutsideClicked = !this.popupRef.current.contains(event.target);

        const wasAnchorClicked =
            this.anchorRef.current && this.anchorRef.current.contains(event.target);
        if (!wasPopupClicked && wasOutsideClicked && !wasAnchorClicked) {
            this.close();
        }
    }

    handleKeyPress(event) {
        if (event.key === ESCAPE_KEY) {
            this.close();
        }
    }

    close() {
        const { isOpened } = this.state;
        if (isOpened === true) {
            this.setState({ isOpened: false });
        }
    }

    createPopperInstance() {
        if (!this.anchorRef.current || !this.popupRef.current) return;
        const { placement } = this.props;
        this.destroyPopperIstance();
        this.popper = new PopperJS(this.anchorRef.current, this.popupRef.current, {
            placement,
        });
    }

    destroyPopperIstance() {
        if (!this.popper) return;
        this.popper.destroy();
        this.popper = undefined;
    }

    renderAnchor() {
        const { anchorRenderer } = this.props;
        const anchorElem = anchorRenderer(this.getArgs());
        return React.cloneElement(anchorElem, { ref: this.anchorRef });
    }

    renderPopup() {
        const { isOpened } = this.state;
        if (!isOpened) return null;
        const { popupRenderer } = this.props;
        const popupElem = popupRenderer(this.getArgs());
        const popupElemWithProps = React.cloneElement(popupElem, {
            ref: this.popupRef,
            'data-popup': 'true',
        });
        return ReactDOM.createPortal(popupElemWithProps, document.body);
    }

    render() {
        return (
            <React.Fragment>
                {this.renderAnchor()}
                {this.renderPopup()}
            </React.Fragment>
        );
    }
}

PopupBase.displayName = 'PopupBase';

PopupBase.propTypes = {
    /**
     * Function, that returns an element that triggers popup.
     * It will be called with a single object as argument that contains:
     *      * setPopupVisibility {function} - can be called with true/false to show/hide the popup
     *      * isOpened {boolean} - the current state of the popup
     * NOTE: The returned element should support refForward, but should not have it set.
     *     If you need to access the ref, pass the ref with anchorRef prop (see below)
     */
    anchorRenderer: PropTypes.func.isRequired,
    /**
     * Function, that returns popup element.
     * It will be called with a single object as argument that contains:
     *      * setPopupVisibility {function} - can be called with true/false to show/hide the popup
     *      * isOpened {boolean} - the current state of the popup
     * NOTE: The returned element should support refForward, but should not have it set.
     *     If you need to access the ref, pass the ref with popupRef prop (see below)
     */
    popupRenderer: PropTypes.func.isRequired,
    /** ref object for anchor */
    anchorRef: PropTypes.object, // eslint-disable-line react/forbid-prop-types
    /** ref object for popup */
    popupRef: PropTypes.object, // eslint-disable-line react/forbid-prop-types
    /** placement of the popup dialog relative to anchor */
    placement: PropTypes.oneOf(POPUP_PLACEMENTS),
};

PopupBase.defaultProps = {
    anchorRef: null,
    popupRef: null,
    placement: 'bottom-start',
};

export default PopupBase;
