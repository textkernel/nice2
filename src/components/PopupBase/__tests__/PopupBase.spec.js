/* eslint-disable react/display-name, react/prop-types */
import React from 'react';
import toJson from 'enzyme-to-json';
import { Button } from '../../..';
import { ESCAPE_KEY } from '../../../constants';
import PopupBase from '../PopupBase';
import PopoverDummy from '../__mocks__/PopoverDummy';

describe('<PopupBase> that adds basic anchor/popup functionality to rendered components', () => {
    const anchorRendererMock = ({ setPopupVisibility, isOpen }) => (
        <Button onClick={() => setPopupVisibility(!isOpen)}>Toggle popup</Button>
    );
    const popupRendererMock = ({ setPopupVisibility }) => (
        <PopoverDummy setPopupVisibility={setPopupVisibility} />
    );
    let wrapper;

    describe('rendering', () => {
        beforeEach(() => {
            wrapper = mount(
                <PopupBase anchorRenderer={anchorRendererMock} popupRenderer={popupRendererMock} />
            );
        });

        it('should render with minimal props correctly', () => {
            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('Popover')).toHaveLength(0);
        });
        it('should render popup when requested', () => {
            // trigger setPopupVisibility(true) through our dummy component
            wrapper.find('button').simulate('click');

            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('Popover')).toHaveLength(1);
        });
        it('should render popup in portal when requested', () => {
            wrapper = mount(
                <PopupBase
                    anchorRenderer={anchorRendererMock}
                    popupRenderer={popupRendererMock}
                    renderInPortal
                />
            );

            // trigger setPopupVisibility(true) through our dummy component
            wrapper.find('button').simulate('click');

            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('Popover')).toHaveLength(1);
            expect(wrapper.find('Portal')).toHaveLength(1);
        });
        it('should close popup when requested', () => {
            // trigger setPopupVisibility(true) through our dummy component
            const triggerButton = wrapper.find('button');
            triggerButton.simulate('click');
            expect(wrapper.find('Popover')).toHaveLength(1);

            triggerButton.simulate('click');
            expect(wrapper.find('Popover')).toHaveLength(0);
        });
    });

    describe('click and keydown event handling', () => {
        let togglePopup;
        const onCloseMock = jest.fn();

        // see: https://medium.com/@DavideRama/testing-global-event-listener-within-a-react-component-b9d661e59953
        const mockDocumentEventListener = {};
        document.addEventListener = jest.fn((event, cb) => {
            mockDocumentEventListener[event] = cb;
        });

        beforeEach(() => {
            wrapper = mount(
                <PopupBase
                    anchorRenderer={anchorRendererMock}
                    popupRenderer={popupRendererMock}
                    onClose={onCloseMock}
                />
            );

            togglePopup = () => {
                wrapper
                    .find('button')
                    .at(0)
                    .simulate('click');
            };
        });

        afterEach(() => {
            onCloseMock.mockReset();
        });

        it('should close open popup if outside is clicked', () => {
            togglePopup();
            expect(wrapper.find('Popover')).toHaveLength(1);

            mockDocumentEventListener.click({ target: document.body });
            wrapper.update();

            expect(wrapper.find('Popover')).toHaveLength(0);
        });
        it('should call onClose if outside is clicked', () => {
            togglePopup();

            mockDocumentEventListener.click({ target: document.body });
            wrapper.update();

            expect(onCloseMock).toHaveBeenCalled();
        });
        it('should not close open popup if popup is clicked', () => {
            togglePopup();
            expect(wrapper.find('Popover')).toHaveLength(1);

            // clicking directly in the element won't trigger global listener, hence we use our magic mock
            mockDocumentEventListener.click({
                target: wrapper
                    .find('Popover')
                    .find('p')
                    .at(0)
                    .getDOMNode(),
            });
            wrapper.update();

            expect(wrapper.find('Popover')).toHaveLength(1);
        });
        it('should not close open popup if button is clicked (ignoring functionality added by the renderer)', () => {
            togglePopup();
            expect(wrapper.find('Popover')).toHaveLength(1);

            // clicking directly in the element won't trigger global listener, hence we use our magic mock
            // this also ensures that event handlers defined by the renderer prop are not triggered.
            mockDocumentEventListener.click({
                target: wrapper
                    .find('button')
                    .at(0)
                    .getDOMNode(),
            });
            wrapper.update();

            expect(wrapper.find('Popover')).toHaveLength(1);
        });
        it('should close open popup on Escape press', () => {
            togglePopup();
            expect(wrapper.find('Popover')).toHaveLength(1);

            mockDocumentEventListener.keydown({ key: ESCAPE_KEY });
            wrapper.update();

            expect(wrapper.find('Popover')).toHaveLength(0);
        });
        it('should call onClose on Escape press', () => {
            togglePopup();

            mockDocumentEventListener.keydown({ key: ESCAPE_KEY });
            wrapper.update();

            expect(onCloseMock).toHaveBeenCalled();
        });
    });
});
