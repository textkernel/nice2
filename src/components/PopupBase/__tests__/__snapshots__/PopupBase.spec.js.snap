// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PopupBase> that adds basic anchor/popup functionality to rendered components rendering should render popup in portal when requested 1`] = `
<PopupBase
  anchorRef={null}
  anchorRenderer={[Function]}
  placement="bottom-start"
  popupRef={null}
  popupRenderer={[Function]}
  renderInPortal={true}
>
  <Button
    onClick={[Function]}
  >
    <button
      className="Button"
      onClick={[Function]}
    >
      Toggle popup
    </button>
  </Button>
  <Portal
    containerInfo={
      <body>
        <div
          style="background-color: rgb(255, 238, 221); padding: 10px; border-radius: 5px; border: 1px solid #ccc; box-shadow: 0 3px 6px 6px rgba(0, 0, 0, 0.05);"
        >
          <p>
            I am the content of pop up.
          </p>
          <p>
            Can be a form or anything.
          </p>
          <button
            class="Button"
          >
            Close me!
          </button>
        </div>
      </body>
    }
  >
    <Popover
      data-popup="true"
      setPopupVisibility={[Function]}
    >
      <div
        style={
          Object {
            "backgroundColor": "#FFEEDD",
            "border": "1px solid #CCC",
            "borderRadius": "5px",
            "boxShadow": "0 3px 6px 6px rgba(0, 0, 0, 0.05)",
            "padding": "10px",
          }
        }
      >
        <p>
          I am the content of pop up.
        </p>
        <p>
          Can be a form or anything.
        </p>
        <Button
          onClick={[Function]}
        >
          <button
            className="Button"
            onClick={[Function]}
          >
            Close me!
          </button>
        </Button>
      </div>
    </Popover>
  </Portal>
</PopupBase>
`;

exports[`<PopupBase> that adds basic anchor/popup functionality to rendered components rendering should render popup when requested 1`] = `
<PopupBase
  anchorRef={null}
  anchorRenderer={[Function]}
  placement="bottom-start"
  popupRef={null}
  popupRenderer={[Function]}
  renderInPortal={false}
>
  <Button
    onClick={[Function]}
  >
    <button
      className="Button"
      onClick={[Function]}
    >
      Toggle popup
    </button>
  </Button>
  <Popover
    data-popup="true"
    setPopupVisibility={[Function]}
  >
    <div
      style={
        Object {
          "backgroundColor": "#FFEEDD",
          "border": "1px solid #CCC",
          "borderRadius": "5px",
          "boxShadow": "0 3px 6px 6px rgba(0, 0, 0, 0.05)",
          "padding": "10px",
        }
      }
    >
      <p>
        I am the content of pop up.
      </p>
      <p>
        Can be a form or anything.
      </p>
      <Button
        onClick={[Function]}
      >
        <button
          className="Button"
          onClick={[Function]}
        >
          Close me!
        </button>
      </Button>
    </div>
  </Popover>
</PopupBase>
`;

exports[`<PopupBase> that adds basic anchor/popup functionality to rendered components rendering should render with minimal props correctly 1`] = `
<PopupBase
  anchorRef={null}
  anchorRenderer={[Function]}
  placement="bottom-start"
  popupRef={null}
  popupRenderer={[Function]}
  renderInPortal={false}
>
  <Button
    onClick={[Function]}
  >
    <button
      className="Button"
      onClick={[Function]}
    >
      Toggle popup
    </button>
  </Button>
</PopupBase>
`;
